### PR TITLE
#1864 - MSFAA Receive File E2E Test Automation

### PR DESCRIPTION
**Pull request by andrewsignori-aot**
- Created the method `mockDownloadFiles` to allow the simulation of a file downloaded from SFTP.
The file will be available on the disk and the SSH client methods `list`  and `get` will be mocked to allow the SSH methods to be invoked as usual but returning the mocked file instead.
- Small change on the logs with the intention to be more clear to read.
- Created the methods `getStructuredRecords` and `createFileFromStructuredRecords` to allow the file content manipulation before the mocked file content is downloaded.
- Created three mock files to be used as downloaded mocks files from SFTP. Some files are reused and have their content changed. The idea of this PR is to show both approaches that can be used.
    - First approach: use a new mocked file when convenient,
    - Second approach: reuse an existing mocked file and adapt its content as needed. Ideally used when a small modification is needed. Up to the developer to decide.

### E2E tests
√ Should process an MSFAA response with confirmations and a cancellation and update all records when the file is received as expected. (280 ms)                                                                                                                           
    √ Should successfully process 2 MSFAA records when a file has 3 records but one throws an error during DB update. (212 ms)
    √ Should throw an error when the MSFAA file contains an invalid SIN hash total. (14 ms)                                             
    √ Should throw an error when the MSFAA file contains an invalid record count. (7 ms)                                                
    √ Should throw an error when the MSFAA file contains an invalid header code. (9 ms)                                                 
    √ Should throw an error when the MSFAA file contains an invalid footer code. (9 ms)

### CODE REVIEW

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/msfaa-integration/_tests_/msfaa-part-time-process-response-integration.scheduler.e2e-spec.ts [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+
+    beforeEach(async () => {
+      jest.clearAllMocks();
+      // Force any not signed MSFAA to be signed to ensure that new
+      // ones created will be the only ones available to be updated.
+      await db.msfaaNumber.update(
+        { dateSigned: IsNull() },
+        { dateSigned: getISODateOnlyString(new Date()) },
+      );
+      // Cancel any pending MSFAA.
```
Comment (ann-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/msfaa-integration/_tests_/msfaa-part-time-process-response-integration.scheduler.e2e-spec.ts:60]: mm...here what does it mean "pending MSFAA"? when `dateSigned === null` then MSFAA is pending, right?

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/msfaa-integration/_tests_/msfaa-part-time-process-response-integration.scheduler.e2e-spec.ts [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+        },
+        where: {
+          id: In(msfaaIDs),
+        },
+        order: {
+          msfaaNumber: "ASC",
+        },
+      });
+      expect(msfaaUpdatedRecords).toHaveLength(msfaaInputData.length);
+      const [fistSignedMSFAA, cancelledMSFAA, secondSignedMSFAA] =
```
Comment (andrepestana-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/msfaa-integration/_tests_/msfaa-part-time-process-response-integration.scheduler.e2e-spec.ts:124]: Minor: the name of the variable should be "firstSignedMSFAA".

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/msfaa-integration/_tests_/msfaa-part-time-process-response-integration.scheduler.e2e-spec.ts [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+        msfaaUpdatedRecords;
+      // Validate fist confirmed record.
+      expect(fistSignedMSFAA.dateSigned).toBe("2021-11-20");
+      expect(fistSignedMSFAA.serviceProviderReceivedDate).toBe("2021-11-21");
+      // Validate cancelled record.
+      expect(cancelledMSFAA.cancelledDate).toBe("2021-11-24");
+      expect(cancelledMSFAA.newIssuingProvince).toBe("ON");
```
Comment (ann-aot) [inline sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/msfaa-integration/_tests_/msfaa-part-time-process-response-integration.scheduler.e2e-spec.ts:131]: just to check `serviceProviderReceivedDate` is not for canceled entry

**Review by andrepestana-aot — APPROVED** [review]
Looks good. Great job!

**Review by ann-aot — APPROVED** [review]
Good work with receive file test👍 just minor comments.

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/msfaa-integration/_tests_/msfaa-part-time-process-response-integration.scheduler.e2e-spec.ts (hunk 3 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/msfaa-integration/_tests_/msfaa-part-time-process-response-integration.scheduler.e2e-spec.ts:131]: Right, `serviceProviderReceivedDate`  is set while confirming an MSFAA.

File: sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/msfaa-integration/_tests_/msfaa-part-time-process-response-integration.scheduler.e2e-spec.ts (hunk 2 above)
Comment (andrewsignori-aot) [reply sources/packages/backend/apps/queue-consumers/src/processors/schedulers/esdc-integration/msfaa-integration/_tests_/msfaa-part-time-process-response-integration.scheduler.e2e-spec.ts]: Fixed.

File: sources/packages/backend/libs/integrations/src/esdc-integration/msfaa-integration/msfaa-response.processing.service.ts [hunk 4]
```diff
@@ -76,7 +83,7 @@ export class MSFAAResponseProcessingService {
       try {
         await this.processCancelledRecord(cancelledRecord);
         result.processSummary.push(
-          `Status cancelled record from line ${cancelledRecord.lineNumber}.`,
+          `Record from line ${cancelledRecord.lineNumber}, updated as canceled.`,
```
Comment (dheepak-aot) [inline sources/packages/backend/libs/integrations/src/esdc-integration/msfaa-integration/msfaa-response.processing.service.ts:86]: cancelled ?

Comment (andrewsignori-aot) [reply sources/packages/backend/libs/integrations/src/esdc-integration/msfaa-integration/msfaa-response.processing.service.ts]: Changed.

File: sources/packages/backend/libs/test-utils/src/mocks/ssh-service-mock.ts [hunk 5]
```diff
@@ -41,3 +53,71 @@ export function getUploadedFile(
...
+  fileTransformation?: (fileContent: string) => string,
+): void {
+  const fileInfos = filePaths.map(
+    (filePath) => ({ name: filePath } as Client.FileInfo),
+  );
+  sshClientMock.list.mockResolvedValue(fileInfos);
+  sshClientMock.get.mockImplementation((filePath: string) => {
+    let fileContent = readFileSync(filePath).toString();
+    if (fileTransformation) {
+      fileContent = fileTransformation(fileContent);
```
Comment (dheepak-aot) [inline sources/packages/backend/libs/test-utils/src/mocks/ssh-service-mock.ts:85]: 👍

Comment (dheepak-aot) [inline sources/packages/backend/libs/test-utils/src/mocks/ssh-service-mock.ts:85]: Thanks for adding this to support the file manipulation. I see that currently we do not have a usage for this but I assume this is going to be helpful when we add more scenarios to the testing (unit tests) right?

**Review by dheepak-aot — APPROVED** [review]
Great work 👍

**Review by guru-aot — APPROVED** [review]
LGTM, nice work @andrewsignori-aot

**Pull request merged**